### PR TITLE
feat: allow path-based branch name

### PR DESCRIPTION
**Pull request by lktslionel**
## Summary

**As A** User 
**I Want to** use path-based branch name as part of the template path
**So that** I could use more branch name pattern to classify my caz templates

## Acceptance criteria

```gherkin
Given the template ref `owner/template#category/subcategory`
When I run `npx caz @owner/template#category/subcategory test`
Then `caz` should successfully download the template and start the scaffolding process
```

### CODE REVIEW

**Comment by lktslionel** [comment]
This feature was required to enable users to use any kind of valid branch name pattern.

File: src/init/resolve.ts [hunk 1]
```diff
@@ -37,7 +37,10 @@ export const getTemplateUrl = async (input: string): Promise<string> => {
   input = input.includes('/') ? input : `${config.official}/${input}`
   input = input.includes('#') ? input : `${input}#${config.branch}`
 
-  const [owner, name, branch] = input.split(/\/|#/)
+  const regex: RegExp = /(?<owner>[\w-]+)\/(?<name>[\w-]+)#(?<branch>[\w-/]+)/mg
```
Comment (zce) [inline src/init/resolve.ts]: Please confirm `mg` regex flags is necessary

File: src/init/resolve.ts [hunk 2]
```diff
@@ -37,7 +37,10 @@ export const getTemplateUrl = async (input: string): Promise<string> => {
   input = input.includes('/') ? input : `${config.official}/${input}`
```
Comment (zce) [inline src/init/resolve.ts]: If there is a slash in the branch name, the judgment here is invalid, like this:

- `foo#bar/baz`

File: src/init/resolve.ts [hunk 3]
```diff
@@ -37,7 +37,10 @@ export const getTemplateUrl = async (input: string): Promise<string> => {
...
   input = input.includes('#') ? input : `${input}#${config.branch}`
 
-  const [owner, name, branch] = input.split(/\/|#/)
+  const regex: RegExp = /(?<owner>[\w-]+)\/(?<name>[\w-]+)#(?<branch>[\w-/]+)/mg
+  const match = regex.exec(input)
+
+  const { owner, name, branch } = ((match?.groups) != null) || {}
```
Comment (zce) [inline src/init/resolve.ts]: Please verify that the code is valid

**Review by zce — COMMENTED** [review]
You need to evaluate the necessity of this path-based branch name

File: src/init/resolve.ts (hunk 1 above)
Comment (lktslionel) [reply src/init/resolve.ts]: I switched to a better implementation

File: src/init/resolve.ts (hunk 2 above)
Comment (lktslionel) [reply src/init/resolve.ts]: The new implementation fixed this issue.

File: src/init/resolve.ts (hunk 3 above)
Comment (lktslionel) [reply src/init/resolve.ts]: I switch to a better implementation.

**Comment by lktslionel** [comment]
@zce 
> You need to evaluate the necessity of this path-based branch name
Thanks for the review. This feature is very useful is you want not to organize your template by branch name.
I plan to structure my templates using the following branch name pattern and it will be so great for `caz` to support this:

The template is named after my project type, structure this way: `<platform>/<type>`.

###### Eg:

* web/app
* mobile/app
* os/cli

I hope it makes sense now 😊.

Regards,

**Comment by zce** [comment]
It seems that you want to manage all your templates through different branches of a single repo. I don't recommend this. I think it's bad for template maintenance.

I suggest you use one organization to manage all your templates, and branches should be used to control different features as much as possible.

**Pull request merged**